File: trie.cpp
#include <iostream>
#include <vector>

struct Leaf {
    std::string key;
    std::string value;
    Leaf* next;
    Leaf(std::string key, std::string value, Leaf* next) {
        this->key = key;
        this->value = value;
        this->next = next;
    }
};

class Trie {
public:
    std::vector<Trie*> connections;
    Leaf* leaf;
    Trie() {
        connections = std::vector<Trie*>(26);
        for (int i = 0; i < 26; i++) {
            connections[i] = NULL;
        }
        leaf = NULL;
    }

    void add(std::string key, std::string value) {
        Trie* start = this;
        for (char value: key) {
            int index = (int)value % 26;
            if (start->connections[index] == NULL) {
                start->connections[index] = new Trie();
            }
            start = start->connections[index];
        }
        start->leaf = new Leaf(key, value, start->leaf);
    }

    std::string get(std::string key) {
        Trie* start = this;
        for (char value: key) {
            int index = (int)value % 26;
            if (start->connections[index] == NULL) {
                return NULL;
            }
            start = start->connections[index];
        }
        Leaf* currLeaf = start->leaf;
        while (currLeaf->key != key && currLeaf->next != NULL)  {
            currLeaf = currLeaf->next;
        }
        return currLeaf->value;
    }
};

int main(void) {
    Trie trie = Trie();
    trie.add("Apples", "5");
    trie.add("Bob", "Gay");
    std::cout << trie.get("Apples") << std::endl;
    std::cout << trie.get("Bob") << std::endl;
    std::cout << trie.get("Alice") << std::endl;
    return 0;
}

File: trie.java
import java.util.ArrayList;

class Leaf {
    public String key;
    public String value;
    public Leaf next;
    public Leaf(String key, String value) {
        this.key = key;
        this.value = value;
        this.next = null;
    }
}

class Trie {
    public char value;
    public ArrayList<Trie> values;
    public Leaf leaf;
    public Trie(char value) {
        this.value = value;
        this.values = new ArrayList<Trie>();
        for (int i = 0; i < 26; i++) {
            this.values.add(i, null);
        }
    }

    public void add(String key, String value) {
        Trie curr = this;
        for (int i = 0; i < key.length(); i++) {
            int index = (int)key.charAt(i) % 26;
            if (curr.values.get(index) == null) {
                curr.values.add(index, new Trie(key.charAt(i)));
            }
            curr = curr.values.get(index);
        }
        Leaf tmp = curr.leaf;
        curr.leaf = new Leaf(key, value);
        curr.leaf.next= tmp;
    }

    public String get(String key) {
        Trie curr = this;
        for (int i = 0; i < key.length(); i++) {
            int index = (int)key.charAt(i) % 26;
            if (curr.values.get(index) == null) {
                return null;
            }
            curr = curr.values.get(index);
        };
        Leaf tmp = curr.leaf;
        while (tmp.key != key && tmp.next != null) {
            tmp = tmp.next;
        }
        return tmp.value;
    }
}

public class Main {
    public static void main(String[] args) {
        Trie root = new Trie(' ');
        root.add("Apples", "Epic");
        root.add("Blitz", "cheese");
        System.out.println(root.get("Apples"));
        System.out.println(root.get("John"));
        System.out.println(root.get("Blitz"));
    }
}
